### PR TITLE
DEPR: deprecate returning a tuple from a callable in iloc indexing

### LINKED ISSUE #53533: BUG: iloc-indexing with callable has inconsistent semantics if the return type is a tuple
### Pandas version checks

- [X] I have checked that this issue has not already been reported.

- [X] I have confirmed this bug exists on the [latest version](https://pandas.pydata.org/docs/whatsnew/index.html) of pandas.

- [ ] I have confirmed this bug exists on the [main branch](https://pandas.pydata.org/docs/dev/getting_started/install.html#installing-the-development-version-of-pandas) of pandas.

### Reproducible Example

```python
import pandas as pd
df = pd.DataFrame({"a": [1, 2, 3, 4], "b": [1, 2, 3, 4]})

scalar_value = df.iloc[(1, 0)] # a single scalar
some_rows = df.iloc[lambda df: (1, 0)] # rows 1 and 0
df.iloc[(1, 0), 0] # indexerror (as expected)
df.iloc[(lambda df: (1, 0)), 0] # indexerror (as expected)
df.iloc[lambda df: ((1, 0), 0)] # ValueError setting an array element with a sequence
```

### Issue Description

The documentation for `iloc` says one possible argument for indexing can be:

> A callable function with one argument (the calling Series or DataFrame) and that returns valid output for indexing (one of the above). This is useful in method chains, when you don’t have a reference to the calling object, but would like to base your selection on some value.

Now, recently, in #47989, the documentation was updated to add a _final_ bullet point (below the just quoted text) to say:

> A tuple of row and column indexes. The tuple elements consist of one of the above inputs, e.g. (0, 1).

A strict reading of this says that the `callable` is not allowed to return a `tuple` (since a `tuple` is not "one of the above").

OTOH, if we accept that callables are allowed in either slot for rows or columns, then the first two examples above should probably behave identically.

What are the intended semantics here?

### Expected Behavior

callables should either in all cases not be allowed to return tuples for indexing, or else should have consistent semantics.

For consistent semantics, one possible mechanism to do this is to have a fixed point normalisation rule:

```
def normalize(key, frame):
    if callable(key):
        return normalize(key(frame), frame)
    if isinstance(tuple, key):
        row, col = key
        return normalize(row, frame), normalize(col, frame)
    return key
```

### Installed Versions

<details>

INSTALLED VERSIONS
------------------
commit           : 37ea63d540fd27274cad6585082c91b1283f963d
python           : 3.11.3.final.0
python-bits      : 64
OS               : Linux
OS-release       : 5.19.0-43-generic
Version          : #44~22.04.1-Ubuntu SMP PREEMPT_DYNAMIC Mon May 22 13:39:36 UTC 2
machine          : x86_64
processor        : x86_64
byteorder        : little
LC_ALL           : None
LANG             : en_GB.UTF-8
LOCALE           : en_GB.UTF-8

pandas           : 2.0.1
numpy            : 1.24.3
pytz             : 2023.3
dateutil         : 2.8.2
setuptools       : 67.7.2
pip              : 23.1.2
Cython           : None
pytest           : None
hypothesis       : None
sphinx           : None
blosc            : None
feather          : None
xlsxwriter       : None
lxml.etree       : None
html5lib         : None
pymysql          : None
psycopg2         : None
jinja2           : None
IPython          : 8.13.2
pandas_datareader: None
bs4              : None
bottleneck       : None
brotli           : None
fastparquet      : None
fsspec           : None
gcsfs            : None
matplotlib       : None
numba            : None
numexpr          : None
odfpy            : None
openpyxl         : None
pandas_gbq       : None
pyarrow          : None
pyreadstat       : None
pyxlsb           : None
s3fs             : None
scipy            : None
snappy           : None
sqlalchemy       : None
tables           : None
tabulate         : None
xarray           : None
xlrd             : None
zstandard        : None
tzdata           : 2023.3
qtpy             : None
pyqt5            : None

</details>

### PR DESCRIPTION
**Pull request by wence-**
The current semantics are that tuple-destructuring of the key is
performed before unwinding any callables. As such, if a callable
returns a tuple for iloc, it will be handled incorrectly. To avoid
this, explicitly deprecate support for this behaviour.

- [x] closes #53533
- [x] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [x] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- ~[ ] Added [type annotations](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#type-hints) to new arguments/methods/functions.~
- [x] Added an entry in the latest `doc/source/whatsnew/vX.X.X.rst` file if fixing a bug or adding a new feature.

### CODE REVIEW

**Comment by wence-** [comment]
This deprecation is as discussed in #53533. TBH, I am not sure that this deprecation is the right approach. I would think it more natural that the order of events in producing "normalized" indices would be:

1. apply callable if appropriate
2. apply tuple destructuring to obtain row and column keys
3. apply normal (non-callable/tuple) indexing rules to the row and column keys

This would have very consistent and easy to understand semantics.

The current approach is much messier:

1. If tuple, apply tuple destructuring, then callable if appropriate, and then normal indexing rules to row and column keys.
2. If _not_ tuple, apply callable to obtain a row (axis-0) key, then apply normal indexing rules (except that the called low-level code doesn't expect tuples so just treats them as list-like).

This PR deprecates (by explicitly warning in case (2)) returning a tuple from a callable. With the intention that it will in the future always raise `IndexError`.

File: doc/source/whatsnew/v2.1.0.rst [hunk 1]
```diff
@@ -298,6 +298,7 @@ Deprecations
 - Deprecated parameter ``obj`` in :meth:`GroupBy.get_group` (:issue:`53545`)
 - Deprecated positional indexing on :class:`Series` with :meth:`Series.__getitem__` and :meth:`Series.__setitem__`, in a future version ``ser[item]`` will *always* interpret ``item`` as a label, not a position (:issue:`50617`)
 - Deprecated strings ``T``, ``t``, ``L`` and ``l`` denoting units in :func:`to_timedelta` (:issue:`52536`)
+- Deprecated support for returning a tuple from a callable in ``iloc``-indexing. Place callables inside a tuple if you need to generate row and column indices using functions (:issue:`53533`)
```
Comment (topper-123) [inline doc/source/whatsnew/v2.1.0.rst:301]: Can you explain what `Place callables inside a tuple if you need to generate row and column indices using functions ` means.

File: pandas/core/indexing.py [hunk 2]
```diff
@@ -162,7 +163,8 @@ def iloc(self) -> _iLocIndexer:
         - A ``callable`` function with one argument (the calling Series or
           DataFrame) and that returns valid output for indexing (one of the above).
           This is useful in method chains, when you don't have a reference to the
-          calling object, but would like to base your selection on some value.
+          calling object, but would like to base your selection on
+          some value. Note that returning a tuple from a callable is deprecated.
```
Comment (topper-123) [inline pandas/core/indexing.py:167]: The deprecation message should be in a separate section starting with `.. deprecated:: 2.1.0`. Look through the code base for examples.

File: pandas/core/indexing.py [hunk 3]
```diff
@@ -1130,6 +1144,17 @@ def __getitem__(self, key):
...
+                self.name == "iloc"
+                and callable(key)
+                and isinstance(maybe_callable, tuple)
+            ):
+                warnings.warn(
+                    "Returning a tuple from a callable in iLocation indexing "
+                    "is deprecated and will be removed in a future version",
+                    FutureWarning,
+                    stacklevel=find_stack_level(),
+                )
```
Comment (topper-123) [inline pandas/core/indexing.py:1157]: This is very similar to the above. Can you refactor them into a method function `apply_to_iloc_if_callable` or similar.

Also the same issue is with `.loc` indexing, isn't it?

File: pandas/tests/frame/indexing/test_indexing.py [hunk 4]
```diff
@@ -1000,6 +1000,14 @@ def test_single_element_ix_dont_upcast(self, float_frame):
         result = df.loc[[0], "b"]
         tm.assert_series_equal(result, expected)
 
+    def test_iloc_callable_tuple_return_value(self):
+        df = DataFrame(np.random.randn(10, 4), index=range(0, 20, 2))
```
Comment (topper-123) [inline pandas/tests/frame/indexing/test_indexing.py:1004]: Add `GH53769` at the top of the test function.

File: pandas/core/indexing.py (hunk 3 above)
Comment (topper-123) [reply pandas/core/indexing.py:1157]: Also, you should be able to return a tuple, if the index is a `MultiIndex` though.

**Review by topper-123 — COMMENTED** [review]
A few comments.

File: pandas/core/indexing.py (hunk 3 above)
Comment (wence-) [reply pandas/core/indexing.py:1157]: > Also the same issue is with .loc indexing, isn't it?

No. A singleton tuple is (potentially) a valid key for `.loc` indexing (although this is also treated inconsistently and generally incorrectly by this code, see https://github.com/pandas-dev/pandas/issues/53535).

File: pandas/tests/frame/indexing/test_indexing.py (hunk 4 above)
Comment (wence-) [reply pandas/tests/frame/indexing/test_indexing.py:1004]: Do you mean the link to this PR, or rather the link to the issue report (53533)?

**Comment by wence-** [comment]
[Aside: I did not write this quoted comment, please don't edit my comments to put words in my mouth]

> Currently callables operate on axis 0 only in loc/iloc. 

This is not true, `LocationIndexer` has this implementation in `__getitem__`:

```python
    def __getitem__(self, key):
        check_dict_or_set_indexers(key)
        if type(key) is tuple:
            key = tuple(list(x) if is_iterator(x) else x for x in key)
            key = tuple(com.apply_if_callable(x, self.obj) for x in key)
            if self._is_scalar_access(key):
                return self.obj._get_value(*key, takeable=self._takeable)
            return self._getitem_tuple(key)
```
So, if the key is a tuple, we destructure, and then apply the callable separately on the destructured axis-wise indexers if necessary.

Witness:
```python
In [2]: df = pd.DataFrame({"a": [1, 2, 3], "b": [1, 2, 3], "c": [2, 3, 4]})

In [3]: df.iloc[[0, 1], lambda df: [1, 2]]
Out[3]: 
   b  c
0  1  2
1  2  3

In [4]: df.loc[[1, 0], lambda df: ["a", "c"]]
Out[4]: 
   a  c
1  2  3
0  1  2

```

File: doc/source/whatsnew/v2.1.0.rst (hunk 1 above)
Comment (wence-) [reply doc/source/whatsnew/v2.1.0.rst:301]: In a single sentence? Probably not. Perhaps I should just link to the indexing documentation instead.

Comment (wence-) [reply doc/source/whatsnew/v2.1.0.rst]: Done.

File: pandas/core/indexing.py (hunk 3 above)
Comment (wence-) [reply pandas/core/indexing.py]: Done (the refactor).

File: pandas/tests/frame/indexing/test_indexing.py (hunk 4 above)
Comment (wence-) [reply pandas/tests/frame/indexing/test_indexing.py:1005]: Done.

**Comment by wence-** [comment]
Ready for another look, thanks

File: pandas/core/indexing.py [hunk 5]
```diff
@@ -1116,6 +1126,17 @@ def _getitem_nested_tuple(self, tup: tuple):
     def _convert_to_indexer(self, key, axis: AxisInt):
         raise AbstractMethodError(self)
 
+    def _check_deprecated_callable_usage(self, key: Any, maybe_callable: T) -> T:
+        # GH53533
+        if self.name == "iloc" and callable(key) and isinstance(maybe_callable, tuple):
+            warnings.warn(
+                "Returning a tuple from a callable in iLocation indexing "
```
Comment (mroeschke) [inline pandas/core/indexing.py:1133]:
```suggestion
                "Returning a tuple from a callable with iloc "
```

Nit

**Review by mroeschke — COMMENTED** [review]
One small nit and a merge conflict otherwise looks good

**Review by mroeschke — APPROVED** [review]

**Comment by mroeschke** [comment]
Thanks @wence-

**Comment by wence-** [comment]
Thanks, and sorry for not doing a better job babysitting this.

**Pull request merged**